---
title: "CCA to correct tissue-by-RP-gene matrix"
#author: "Charleen D. Adams"
date: "April 28, 2020"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
**Author: Charleen D. Adams**

## Abbreviations and definitions
 - Gene-Tissue Expression (GTex) Project -- database of tissue-specific gene expression and SNPs associated with gene expression
 - Ribosomal protein (RP)
 - expression quantitative trait loci (eQTL) -- SNPs associated with gene expression
 - significance threshold (*Q*-value for 'eGene' status <0.05 & nominal *P*<5x10-6) -- analyses here use the set of GTex eQTLs meet this criteria
 - Canonical Correlation Analysis (CCA)

## Overview
I had made a **correlation matrix** of the RP eQTLs across 49 GTex tissues. However, it is likely that the correlations are confounded by a factor related to which donor donated which tissues. 

## Correlation matrix of tissues by RP eQTLs across the RP genes
```{r echo=FALSE, out.width="100%", fig.align='center'}
knitr::include_graphics("C:/Users/charl/Dropbox/Harvard/ribosomal_proteins/Main_Tables_Paper/v8/old_heatmap.png")
```

To investigate this possibility and correct for the influence of latent factors related to donor contribution, I used CCA. CCA is a tool that is similar to multivariate multiple regression except it also considers dimensionality, operating on a principle similar to prinicpal components analysis. 

```{r cars, messages=FALSE, warning=FALSE, include=FALSE}
setwd('C:/Users/charl/Dropbox/Harvard/ribosomal_proteins/Main_Tables_Paper/v8/')
library(ggplot2)
library(GGally)
library(CCA)
library(lme4)
library(CCP)
library(dplyr)
library(tidyr)
library(EnsDb.Hsapiens.v79) 
library(biomaRt)
library(SNPlocs.Hsapiens.dbSNP.20120608)
library(data.table)
library(Cairo)
library(devtools)
library(MRInstruments) 
library(TwoSampleMR)
library(RadialMR)
library(data.table)
library(tidyverse)
library(RColorBrewer)
library(gplots)
library(tidyverse)
library(packcircles)
library(ggplot2)
library(viridis)
library(igraph)
library(rsconnect)
library(DT)
source("https://neuroconductor.org/neurocLite.R")
#neuro_install('gganatogram')
library(gganatogram)
library(gridExtra)
```

## Read-in the GTex sample annotations 

```{r pressure, echo=FALSE}
#### Function to read in the GTex samples data from a url

read.url <- function(url, ...){
  tmpFile <- tempfile()
  download.file(url, destfile = tmpFile, method = "curl")
  url.data <- fread(tmpFile, ...)
  return(url.data)
}

samples_v8=read.url("https://storage.googleapis.com/gtex_analysis_v8/annotations/GTEx_Analysis_v8_Annotations_SampleAttributesDS.txt")
#write.csv(samples_v8, 'samples_v8.csv')

#### The 'SMTSD' variable stands for 'Tissue Site Detail field' & contains the counts of samples per tissue.
freq=table(samples_v8$SMTSD)
#write.csv(freq, 'v8_tissue_frequencies.csv')
```

## Get the donor IDs
`SAMPID` is the sample ID variable. It contains the donor ID, the tissue site, and aliquot ID. I separated these into variables to obtain donor IDs.

```{r, warning=FALSE}
#### Separate SAMPID into subject, tissue, and aliquot
df <- data.frame(x = samples_v8$SAMPID)
samples_v83=df %>% separate(x, c("GTex","Donor_ID", "tissue_site_ID", "SM","aliquot_ID")) #name choice 'samples_v83' is arbitrary

#### cbind the separated id data with the samples_v8 data
samples_v8_final <- cbind(samples_v8, samples_v83)
donor_tissue=table(samples_v8_final$Donor_ID, samples_v8_final$SMTSD)
#write.csv(donor_tissue, 'donor_tissue.csv')

```

## Count of tissues sites per donor

```{r, warning=FALSE}
#### Determine how many tissue site each donor contributed to
donor_tissue2=read.csv("donor_tissue2.csv")
rownames(donor_tissue2)=donor_tissue2$Donor
#donor_tissue2$Donor <- NULL
cols=colnames(donor_tissue2[2:56])
donor_tissue2$new_col <- apply(donor_tissue2[,cols],1,sum)
donor_tissue2 <- donor_tissue2[order(donor_tissue2$new_col),]
summary(donor_tissue2$new_col)

```

## Remove tissues for which there are no eQTLs

```{r, warning=FALSE}
my_drop=c('Bladder', "Cells - Leukemia cell line (CML)", "Cervix - Ectocervix",
          "Cervix - Endocervix", "Fallopian Tube", "Kidney - Medulla")
samples_v8_final_clean <- samples_v8_final[!which(samples_v8_final$SMTSD %in% my_drop),]
samples_v8_final_clean_tab=table(samples_v8_final_clean$SMTSD)
#write.csv(samples_v8_final_clean_tab, 'samples_v8_final_clean_tab.csv')
```

## Show the tissue-by-donor ID counts
```{r}
donor_tissue_clean=table(samples_v8_final_clean$SMTSD, samples_v8_final_clean$Donor_ID)
```

## Create a correlation matrix for tissue by donor ID

```{r}
#### Get the correlation matrix
donor_tissue_clean_mat<- cor(t(donor_tissue_clean))
Y=donor_tissue_clean_mat


```

## Read in the correlation matrix for tissue by presence of RP eQTLs

```{r, warning=FALSE}
#### Read in the RP eQTL data to tabulate tissue by RP gene 
load('./V8_master_EAF.Rdata')
tissue_gene=table(merged_EAF$tissue,merged_EAF$gene_name )

#### Get the correlation matrix
tissue_gene_mat<- cor(t(tissue_gene))
X=tissue_gene_mat

#### Save the RP lists, latest versions of the dataset (merged_EAF) and the matrices
save(ao, read.url,myvars, myvars_40S,myvars_40S_60S,myvars_60S, myvars_all_mito,myvars_mito_39S, master_v8_n, RP_sig, stringent_RP_sig_beta, merged_EAF, tissue_gene, X, tissue_gene_mat, donor_tissue_clean_mat, Y,  file="./V8_masterRP_lists_mergedEAF_matrices.Rdata")
```
## Construct a matrix of the matrices

$cor(X,Y)$ is a $2\times 2$ matrix, where $X$ is the tissue-RP-gene matrix, $Y$ is the tissue-by-donor matrix, and:    
\[
cor(X,Y) = \begin{bmatrix}
\sum{X X} & \sum{X Y} \\
\sum{Y X} & \sum{Y Y} 
\end{bmatrix}
\]   


```{r}
correl <- matcor(X, Y )
#str(correl)
img.matcor(correl, type = 2)
#dev.off()

```

```{r, include=FALSE}
round(correl$XYcor, 4)
```
## Canonical correlations

```{r}
can_cor1=cc(X,Y)
```

## Raw canonical coefficients
The interpretation of the coefficients is similar to that of multiple regression. If I'm interpretting this correctly, they are the tissue-pair coefficients when holding all the other variables constant; i.e., for the $X$ set, they are the relationship between tissue pairs when accounting for latent structure from donor sample imbalances and all other pairs.
```{r, include=FALSE}
can_cor1[3:4]
#can_cor1_df=can_cor1[3:4]

```

```{r, include=FALSE}
## Compute the canonical loadings
can_cor2=comput(X,Y,can_cor1)
#can_cor2[3:6] #displays the canonical loadings
```

## Standardized coefficients

```{r}
#standardizing the first set of canonical coefficients(X)
 std_coef1<-diag(sqrt(diag(cov(X))))
```
```{r, include=FALSE}
cor_correct=std_coef1%*%can_cor1$xcoef
str(cor_correct)
cor_correct=as.data.frame(cor_correct)
rownames(cor_correct)=rownames(X)
colnames(cor_correct)=colnames(X)
cor_correct=round(cor_correct, 1)
```

## Table of corrected correlations 
```{r, warning=FALSE}
cor_correct_mat<- cor(cor_correct)
cor_correct_round=round(cor_correct_mat, 1)
datatable(cor_correct_round, options = list(pageLength = 5))
```

```{r, include=FALSE}
par(mar=c(7,4,4,2)+0.1) 
png(filename='heatmap.png', width=1800, height=1800)
coul <- colorRampPalette(brewer.pal(8, "PiYG"))(25)
heatmap.2(cor_correct_mat, 
          col=coul, 
          main="CCA-corrected between-tissue correlation of RP eQTLs by gene",
          distfun=function(x) dist(x, method="euclidean"), 
          hclustfun=function(x) hclust(x, method="ward.D2"),  
          cexRow = 1.5, cexCol = 1.5, margins = c(15,15))#scale = "row")
graphics.off()

```
## CCA-corrected heatmap
```{r echo=FALSE, out.width="100%", fig.align='center'}
knitr::include_graphics('C:/Users/charl/Dropbox/Harvard/ribosomal_proteins/Main_Tables_Paper/v8/heatmap.png')
```

## Thoughts
 - I do not know if this approach is correct (nor if I did it right), as I have never done it before. I aimed to correct the tissue-by-RP-gene correlation matrix for likely confounding due to imbalance in how certain donors contributed disproportionately. 
 
 - Assuming I did it correctly, the new data display corrected correlations between the tissues across the RP genes. There doesn't seem to be an obvious pattern between tissues, but maybe that makes sense because each new tissue-pair estimate takes into account all the pairs and the latent influence of the donors.
 
 - So, not sure we can use this analysis, but I wanted to share my thought process. Not sure if it would be worth consulting with a statistician to go over it. If so, who? 

 - Also, the GTex documentation explains that some of the brain tissues (specifically, "Brain - Frontal Cortex" and "Brain- cortex", and "Brain - Cerebellum" and "Brain - Cerebellar Hemisphere") are duplicates. I've kept them in the data for now, but eventually, it would make sense to remove duplicates.
 
## 47 tissues (excluding duplicates)
```{r echo=FALSE, out.width="100%", fig.align='center'}
knitr::include_graphics("C:/Users/charl/Dropbox/Harvard/ribosomal_proteins/Main_Tables_Paper/v8/body_labeled.png")
```
*Body made with `gganatogram`

## Sources 
 
 [*Link*](https://medium.com/analytics-vidhya/canonical-correlation-analysis-cca-in-r-a-non-technical-primer-b67d9bdeb9dd) to the main tutorial I used to teach myself CCA. 
 
 [*Link*](https://www.karlin.mff.cuni.cz/~maciak/NMST539/cvicenie11.html) for the code for intra- and inter-correlations of the matrices.
 
 [*Another*](https://stats.idre.ucla.edu/r/dae/canonical-correlation-analysis/) helpful link to read up on CCA. 
 
## Github
Copy this link into your browser (*Github* refuses to link): https://github.com/charleendadams/ribosomal_protein_eQTLs

 - This markdown file is stored there. 
